File: 0046-permutations/0046-permutations.java
class Solution {
    private void f(int idx,int[] nums,List<List<Integer>>ans){
        if(idx==nums.length){
            List<Integer> ds=new ArrayList<>();
            for(int i=0;i<nums.length;i++){
                ds.add(nums[i]);
            }
            ans.add(ds);
            return;
        }
        for(int i=idx;i<nums.length;i++){
            swap(i,idx,nums);
            f(idx+1,nums,ans);
            swap(i,idx,nums);
        }
    }
    private void swap(int i,int j,int[] nums){
        int t=nums[i];
        nums[i]=nums[j];
        nums[j]=t;
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>>ans=new ArrayList<>();
        f(0,nums,ans);
        return ans;
    }
}

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    void f(int idx,vector<int>&nums,vector<vector<int>>&ans){
        if(idx==nums.size()-1){
            ans.push_back(nums);
        }
        for(int i=idx;i<nums.size();i++){
            swap(nums[idx],nums[i]);
            f(idx+1,nums,ans);
            swap(nums[idx],nums[i]);
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>>ans;
        f(0,nums,ans);
        return ans;
    }
};